### PR TITLE
Copy GLXFBConfig from shared context.

### LINKED ISSUE #1453: Using Filament with shared context exits with BadMatch on Linux
**Describe the bug**

I'm embedding Filament into an GLFW based application on Linux (*Linux arch680 5.2.1-arch1-1-ARCH #1 SMP PREEMPT Sun Jul 14 14:52:52 UTC 2019 x86_64 GNU/Linux*). When sharing the GL context that's created by GLFW the application exists with a *BadMatch* error, [see this output](https://gist.github.com/roxlu/c282d642c353ce96ef19b6359c741bcb#file-log-sh).

**To Reproduce**

I should create a test repository... hopefully I've got time to do that, but for now:

1. Compile filament as described in the readme.
2. Create a simple GLFW app.
3. Initialize filament [like in this test app](https://gist.github.com/roxlu/c282d642c353ce96ef19b6359c741bcb#file-test-filament-with-glfw-cpp).
4. Run the application.

**Expected behavior**

A GLFW app that renders a triangle via Filament. I'm still diving into the code / documentation on how to correctly initialize and use filament in GLFW. [I based my code on this example](https://github.com/cgmb/hello-filament/blob/master/main.cxx ).

**Desktop (please complete the following information):**

 - OS: Arch Linux
 - GPU:  NVIDIA GTX 960
 - Backend: OpenGL

**Additional Info**

I think the BadMatch is caused because the GL context is created with a GLXFBConfig that doesn't match with the GLXFBConfig of the shared GL context. [I've updated PlatformGLX.cpp ](https://gist.github.com/roxlu/c282d642c353ce96ef19b6359c741bcb#file-platformglx-fix-cpp-L115) in such a way that it uses the GLXFBConfig from the shared GL context when it was given by the user.

### PR DESCRIPTION
**Pull request by roxlu**
Fixes #1453

### CODE REVIEW

File: filament/backend/src/opengl/PlatformGLX.cpp [hunk 1]
```diff
@@ -110,14 +121,62 @@ Driver* PlatformGLX::createDriver(void* const sharedGLContext) noexcept {
...
+      // Create a context
+      static int attribs[] = { GLX_DOUBLEBUFFER, True, None };
 
-    int config_count = 0;
-    mGLXConfig = g_glx.chooseFbConfig(mGLXDisplay, DefaultScreen(mGLXDisplay),
-            attribs, &config_count);
-    if (mGLXConfig == nullptr || config_count == 0) {
+      int config_count = 0;
+      mGLXConfig = g_glx.chooseFbConfig(mGLXDisplay, DefaultScreen(mGLXDisplay),
+                                        attribs, &config_count);
```
Comment (romainguy) [inline filament/backend/src/opengl/PlatformGLX.cpp]: Continuation indents are 8 spaces in our code base, we don't align parameters in this way

File: filament/backend/src/opengl/PlatformGLX.cpp [hunk 2]
```diff
@@ -110,14 +121,62 @@ Driver* PlatformGLX::createDriver(void* const sharedGLContext) noexcept {
...
+
+        if (fb_index < 0) {
+            printf("Error: failed to find an `GLXFBConfig` with the requested ID. (exiting).\n");
+            return nullptr;
+        }
+      
+        mGLXConfig = fb_configs + fb_index;
+    }
+
+    if (nullptr == sharedGLContext) {
```
Comment (romainguy) [inline filament/backend/src/opengl/PlatformGLX.cpp]: The previous `if` already tests this, use an `else`

File: filament/backend/src/opengl/PlatformGLX.cpp [hunk 3]
```diff
@@ -110,14 +121,62 @@ Driver* PlatformGLX::createDriver(void* const sharedGLContext) noexcept {
...
-    static int attribs[] = { GLX_DOUBLEBUFFER, True, None };
+    if (nullptr != sharedGLContext) {
+      
+        int r = -1;
+        int used_fb_id = -1;
+        GLXContext shared_ctx = (GLXContext)((void*)sharedGLContext);
+      
+        r = g_glx.queryContext(mGLXDisplay, shared_ctx, GLX_FBCONFIG_ID, &used_fb_id);
+        if (0 != r) {
+            printf("Error: failed to get GLX_FBCONFIG_ID from shared GL context.");
```
[the remark this answers is not in the document]
Comment (roxlu) [reply filament/backend/src/opengl/PlatformGLX.cpp]: Thanks, changed this. The first check on `if(mGLXDisplay == nullptr)` from the existing code uses `printf()`. I've changed this into `utils::slog.e` too.

File: filament/backend/src/opengl/PlatformGLX.cpp [hunk 4]
```diff
@@ -46,6 +46,11 @@ struct GLXFunctions {
     PFNGLXCREATEPBUFFERPROC createPbuffer;
     PFNGLXDESTROYPBUFFERPROC destroyPbuffer;
     PFNGLXMAKECONTEXTCURRENTPROC setCurrentContext;
+  
+    PFNGLXQUERYCONTEXTPROC queryContext; /* When creating a shared GL context, we query the used GLX_FBCONFIG_ID to make sure our display framebuffer attributes match; otherwise making our context current results in a BadMatch https://gist.github.com/roxlu/c282d642c353ce96ef19b6359c741bcb */
```
[the remark this answers is not in the document]
Comment (roxlu) [reply filament/backend/src/opengl/PlatformGLX.cpp]: Done.

File: filament/backend/src/opengl/PlatformGLX.cpp [hunk 5]
```diff
@@ -110,14 +121,62 @@ Driver* PlatformGLX::createDriver(void* const sharedGLContext) noexcept {
         exit(1);
     }
 
-    // Create a context
-    static int attribs[] = { GLX_DOUBLEBUFFER, True, None };
+    if (nullptr != sharedGLContext) {
+
```
[the remark this answers is not in the document]
Comment (roxlu) [reply filament/backend/src/opengl/PlatformGLX.cpp]: Done.

File: filament/backend/src/opengl/PlatformGLX.cpp [hunk 6]
```diff
@@ -110,14 +121,62 @@ Driver* PlatformGLX::createDriver(void* const sharedGLContext) noexcept {
         exit(1);
     }
 
-    // Create a context
-    static int attribs[] = { GLX_DOUBLEBUFFER, True, None };
+    if (nullptr != sharedGLContext) {
+      
+        int r = -1;
+        int used_fb_id = -1;
```
[the remark this answers is not in the document]
Comment (roxlu) [reply filament/backend/src/opengl/PlatformGLX.cpp]: Done.

File: filament/backend/src/opengl/PlatformGLX.cpp [hunk 7]
```diff
@@ -110,14 +121,62 @@ Driver* PlatformGLX::createDriver(void* const sharedGLContext) noexcept {
...
+        GLXContext shared_ctx = (GLXContext)((void*)sharedGLContext);
+      
+        r = g_glx.queryContext(mGLXDisplay, shared_ctx, GLX_FBCONFIG_ID, &used_fb_id);
+        if (0 != r) {
```
[the remark this answers is not in the document]
Comment (roxlu) [reply filament/backend/src/opengl/PlatformGLX.cpp]: Done.

File: filament/backend/src/opengl/PlatformGLX.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (roxlu) [reply filament/backend/src/opengl/PlatformGLX.cpp]: Done.

File: filament/backend/src/opengl/PlatformGLX.cpp [hunk 8]
```diff
@@ -110,14 +121,62 @@ Driver* PlatformGLX::createDriver(void* const sharedGLContext) noexcept {
...
+            printf("Error: failed to get GLX_FBCONFIG_ID from shared GL context.");
+            return nullptr;
+        }
+    
+        int num_configs = 0;
+        GLXFBConfig* fb_configs = g_glx.getFbConfigs(mGLXDisplay, 0, &num_configs);
+
+        if (nullptr == fb_configs) {
```
[the remark this answers is not in the document]
Comment (roxlu) [reply filament/backend/src/opengl/PlatformGLX.cpp]: Done.

File: filament/backend/src/opengl/PlatformGLX.cpp [hunk 9]
```diff
@@ -110,14 +121,62 @@ Driver* PlatformGLX::createDriver(void* const sharedGLContext) noexcept {
...
+            return nullptr;
+        }
+
+        int fb_id = 0;
+        int fb_index = -1;
+    
+        for (int i = 0; i < num_configs; ++i) {
+        
+            r = g_glx.getFbConfigAttrib(mGLXDisplay, fb_configs[i], GLX_FBCONFIG_ID, &fb_id);
+            if (0 != r) {
```
[the remark this answers is not in the document]
Comment (roxlu) [reply filament/backend/src/opengl/PlatformGLX.cpp]: Done.

File: filament/backend/src/opengl/PlatformGLX.cpp (hunk 1 above)
Comment (roxlu) [reply filament/backend/src/opengl/PlatformGLX.cpp]: Done.

File: filament/backend/src/opengl/PlatformGLX.cpp (hunk 2 above)
Comment (roxlu) [reply filament/backend/src/opengl/PlatformGLX.cpp]: Done.

**Comment by romainguy** [comment]
I'll wait for the CI builds to turn green before merging this.

**Pull request merged**